repl: create history file with mode 0600.

XeCycle opened the pull request:
Test code mostly written by Trott
https://github.com/nodejs/node/issues/3392#issuecomment-148634126.

targos commented on lib/internal/repl.js:
```diff
@@ -88,7 +88,7 @@ function setupHistory(repl, historyPath, oldHistoryPath, ready) {
   var pending = false;
   repl.pause();
-  fs.open(historyPath, 'a+', oninit);
+  fs.open(historyPath, 'a+', /* 0600 */ 384, oninit);
```
you can use an octal literal: `0o0600`

Trott commented:
LGTM if CI is happy.

Trott commented:
Fixes: #3392

Trott commented:
CI: https://ci.nodejs.org/job/node-test-commit/859/

Trott commented:
Test fails on Windows: https://ci.nodejs.org/job/node-test-binary-windows/14/RUN_SUBSET=0,VS_VERSION=vs2015,label=win10/tapTestReport/test.tap-155/

XeCycle commented:
Tried it on Windows, and the returned stat is 010666.  I guess that's because Windows uses a different permission model; do you think it okay to skip this test on Windows?

rvagg commented:
I think so but lets ask @nodejs/platform-windows

Trott commented:
I changed 

      assert.strictEqual(mode, '0600', 'REPL history file should be mode 0600');

To just this

      assert.strictEqual(mode, '0600', 'REPL history file should be mode 0600');

So that it would report the actual value. Result is [here](https://ci.nodejs.org/job/node-test-binary-windows/18/RUN_SUBSET=0,VS_VERSION=vs2015,label=win2012r2/tapTestReport/test.tap-155/).

So, Windows is creating the file with mode 0666 rather than 0600.

The question is: Is this a bug in the code that creates the file or is it a limitation of running the Node.js REPL on the Windows operating system?

Trott commented:
The code uses `fs.open()` to set the mode on file creation, so I guess if there's stuff that does not completely work there, a Windows `libuv` person would know the deal?

Trott commented:
[This comment in the `libuv` source](https://github.com/nodejs/node/blob/0803962860ae4cd2aa7355cbcfff3432e430e496/deps/uv/src/win/fs.c#L1065-L1066) would seem to be more-or-less a smoking gun. I think you can just check `common.isWindows()` at the top of the test, and if it is Windows, skip the test. You can emulate other tests like this from `test-process-remove-all-signal-listeners`:

```
if (common.isWindows) {
  console.log('1..0 # Skipped: Win32 doesn\'t have signals, just a kind of ' +
              'emulation, insufficient for this test to apply.');
  return;
}
```

XeCycle commented:
From [docs of CreateFile](https://msdn.microsoft.com/en-us/library/windows/desktop/aa363858(v=vs.85).aspx):

> - The access control lists (ACL) in the default security descriptor for a file or directory are inherited from its parent directory.

So privacy on the history file relies on the user having a fine ACL in his/her home directory (which appers to be like `C:\Users\username`).

Trott commented:
Cool. LGTM if CI passes.

CI: https://ci.nodejs.org/job/node-test-commit/865/

Trott commented:
All test failures are just the unrelated usual suspects and something weird with OS X such that it won't even build or something.

In other words: :+1:

Hopefully someone else can give it an LGTM too.

thefourtheye commented on test/parallel/test-repl-history-perm.js:
```diff
@@ -0,0 +1,52 @@
+'use strict';
+// Flags: --expose_internals
```
Why do we need this?

thefourtheye commented on test/parallel/test-repl-history-perm.js:
```diff
@@ -0,0 +1,52 @@
+
+const checkResults = common.mustCall(function(err, r) {
+  if (err)
+    throw err;
```
You can use `assert.ifError` instead of this `if` block.

XeCycle replied:
I see there are other tests writing `if (err) throw err;` e.g. test-fs-write.  Does the current style explicitly prefer `assert.ifError`?  If not I think this `if` block is also okay.

Fishrock123 replied:
No need to use assert here, this is common practice.

Fishrock123 commented on test/parallel/test-repl-history-perm.js:
```diff
@@ -0,0 +1,52 @@
+});
+
+repl.createInternalRepl(
+  {NODE_REPL_HISTORY: replHistoryPath},
```
Does this pass `make jslint`?

Trott replied:
Yes: https://ci.nodejs.org/job/node-test-linter/23/ (What's the linting concern? Maybe we need a rule tweak?)

Fishrock123 replied:
@Trott no spaces at the edges of object literals? i.e. `{ NODE_REPL_HISTORY: replHistoryPath }`

targos replied:
the related rule (http://eslint.org/docs/rules/space-in-brackets) is not enabled.
Seems that the most common style in the code base is to have no space:
* Rule enabled, forcing spaces: 8451 errors
* Rule enabled, forcing no space: 1674 errors

Fishrock123 commented:
Seems fine to me.

Trott commented:
@Fishrock123 Does that constitute an LGTM?

Fishrock123 commented on test/parallel/test-repl-history-perm.js:
```diff
@@ -0,0 +1,52 @@
+stream.readable = stream.writable = true;
+
+common.refreshTmpDir();
+const replHistoryPath = path.join(common.tmpDir, 'repl_history');
```
Could you make this consistent with [the other test](https://github.com/nodejs/node/blob/master/test/sequential/test-repl-persistent-history.js#L67-L73)? `'.node_repl_history'`

Fishrock123 commented on test/parallel/test-repl-history-perm.js:
```diff
@@ -0,0 +1,52 @@
+  r.input.end();
+  const stat = fs.statSync(replHistoryPath);
+  const mode = '0' + (stat.mode & parseInt('777', 8)).toString(8);
+  assert.strictEqual(mode, '0600', 'REPL history file should be mode 0600');
```
Seems a bit complex, couldn't we just check `0o0600`?

Trott replied:
Since `stat.mode` comes back as something like  `0o100600` you'd still need to do masking:

    assert.strictEqual(stat.mode & 0o777, 0o600);

If you're goal is legibility, you may be best off converting to a string:

    const mode = stat.mode.toString(8);
    assert(mode.endsWith('600'));

jasnell replied:
Personally I prefer the masked version. I agree that this test is too complicated as is.

XeCycle commented:
@Fishrock123 file name changed, not rebasing.

XeCycle commented:
Please check [this commit](https://github.com/XeCycle/node/commit/15327a5123a2d09be99e124406fb8b79bf029df7).  Discovered we have `-i/--interactive`, so tried testing with that, but it fails because of another check [here](https://github.com/nodejs/node/blob/master/lib/internal/repl.js#L64) and [here](https://github.com/nodejs/node/blob/master/lib/internal/repl.js#L30).  Is it intended that a forced interactive session without terminal does not create a history file?

Fishrock123 commented:
> Discovered we have `-i/--interactive`, so tried testing with that, but it fails because of another check [here](https://github.com/nodejs/node/blob/master/lib/internal/repl.js#L64) and [here](https://github.com/nodejs/node/blob/master/lib/internal/repl.js#L30).  Is it intended that a forced interactive session without terminal does not create a history file?

Undefined. `-i/--interactive` should act just like the regular repl as much as possible though imo.

jasnell commented:
What's the status on this one?

XeCycle commented:
Seems like maintainers did not agree on the tests.  Anyway there are no conflicts, a rebase is easy, so let me know if you want this landed.

jasnell commented:
@nodejs/collaborators @nodejs/ctc ... does anyone have any further thoughts on this one?
@XeCycle ... if you'd like to rebase I'll see if I can get an updated review.

XeCycle commented:
@jasnell rebase done.  Still the original test, testing on `createInternalRepl`.

Trott commented:
@nodejs/testing

Fishrock123 commented:
ci: https://ci.nodejs.org/job/node-test-pull-request/2463/

Fishrock123 commented:
ci is green, I can't really make an informative opinion about this though

silverwind commented:
`600` is the right mode for history files, it protects the content from other users in case the home directory is readable. The test seems a bit overly complicated, but overall LGTM.

Fishrock123 commented:
Ok, LGTM then

santigimeno commented:
LGTM

jasnell commented on lib/internal/repl.js:
```diff
@@ -93,7 +93,7 @@ function setupHistory(repl, historyPath, oldHistoryPath, ready) {
   var pending = false;
   repl.pause();
-  fs.open(historyPath, 'a+', oninit);
+  fs.open(historyPath, 'a+', 0o0600, oninit);
```
Nit: Would you find adding a quick code comment with a quick explanation here. Nothing complicated, just enough for someone scanning through to know why the flag was set to this.

jasnell commented:
Added a couple of comments that I'd like to see addressed before this lands. Otherwise LGTM

jasnell commented:
What semver-iness would this be? I'm thinking just a patch (because I'd consider this a bug fix) but figured I'd ask.

Fishrock123 commented:
I don't think it would be major.

XeCycle commented:
@jasnell

> Would you mind adding a quick code comment with a quick explanation here

Added, pointing to this PR and issue.

> Personally I prefer the masked version

Me too, changed it to the masked version.

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/2491/

jasnell commented:
LGTM if CI is green

silverwind commented:
Thanks you! Landed in https://github.com/nodejs/node/commit/b72d048a743dc7dbe8a1c6d9843d358cb55a1c46.
